File: src/test/java/test/MappingTest.java
package test;

import com.github.t1.yaml.model.Document;
import com.github.t1.yaml.model.Mapping;
import com.github.t1.yaml.model.Mapping.Entry;
import com.github.t1.yaml.model.Scalar;
import com.github.t1.yaml.model.Scalar.Line;
import com.github.t1.yaml.model.Sequence;
import com.github.t1.yaml.model.Sequence.Item;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Nested;

class MappingTest extends AbstractYamlTest {
    @Nested class givenBlockMapping extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: blue\nsea: green";
            expected = new Document().node(new Mapping()
                .entry("sky", "blue")
                .entry("sea", "green")
            );
        }
    }

    @Nested class givenBlockMappingWithSpaceInKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky high: blue\nsea deep: green";
            expected = new Document().node(new Mapping()
                .entry("sky high", "blue")
                .entry("sea deep", "green")
            );
        }
    }

    @Nested class givenBlockMappingWithColonInKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:high: blue\nsea:deep: green";
            expected = new Document().node(new Mapping()
                .entry("sky:high", "blue")
                .entry("sea:deep", "green")
            );
        }
    }

    @Nested class givenBlockMappingWithTwoLineValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: high\n" +
                "  blue\n" +
                "sea: low\n" +
                "  green";
            expected = new Document().node(new Mapping()
                .entry("sky", new Scalar().line("high").line("blue"))
                .entry("sea", new Scalar().line("low").line("green"))
            );
        }
    }

    @Nested class givenBlockMappingWithDoubleQuotedKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "\"sky\": blue\n\"sea\": green";
            expected = new Document().node(new Mapping()
                .entry(new Scalar().doubleQuoted().line("sky"), "blue")
                .entry(new Scalar().doubleQuoted().line("sea"), "green")
            );
        }
    }

    @Nested class givenBlockMappingWithDoubleQuotedValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: \"blue\"\nsea: \"green\"";
            expected = new Document().node(new Mapping()
                .entry("sky", new Scalar().doubleQuoted().line("blue"))
                .entry("sea", new Scalar().doubleQuoted().line("green"))
            );
        }
    }

    @Nested class givenBlockMappingWithDoubleQuotedKeyAndValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "\"sky\": \"blue\"\n\"sea\": \"green\"";
            expected = new Document().node(new Mapping()
                .entry(new Scalar().doubleQuoted().line("sky"), new Scalar().doubleQuoted().line("blue"))
                .entry(new Scalar().doubleQuoted().line("sea"), new Scalar().doubleQuoted().line("green"))
            );
        }
    }

    @Disabled
    @Nested class givenBlockMappingWithTwoLineSingleQuotedValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: 'high\n" +
                "blue'\n" +
                "sea: 'low\n" +
                "green'";
            expected = new Document().node(new Mapping()
                .entry("sky", new Scalar().singleQuoted().line("high").line("blue"))
                .entry("sea", new Scalar().singleQuoted().line("low").line("green"))
            );
        }
    }

    @Nested class givenBlockMappingWithSingleQuotedKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "'sky': blue\n'sea': green";
            expected = new Document().node(new Mapping()
                .entry(new Scalar().singleQuoted().line("sky"), "blue")
                .entry(new Scalar().singleQuoted().line("sea"), "green")
            );
        }
    }

    @Nested class givenBlockMappingWithSingleQuotedValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: 'blue'\nsea: 'green'";
            expected = new Document().node(new Mapping()
                .entry("sky", new Scalar().singleQuoted().line("blue"))
                .entry("sea", new Scalar().singleQuoted().line("green"))
            );
        }
    }

    @Nested class givenBlockMappingWithTwoLineDoubleQuotedValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: \"high\n" +
                "blue\"\n" +
                "sea: \"low\n" +
                "green\"";
            expected = new Document().node(new Mapping()
                .entry("sky", new Scalar().doubleQuoted().line("high\nblue"))
                .entry("sea", new Scalar().doubleQuoted().line("low\ngreen"))
            );
        }
    }

    @Nested class givenBlockMappingWithSingleQuotedKeyAndValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "'sky': 'blue'\n'sea': 'green'";
            expected = new Document().node(new Mapping()
                .entry(new Scalar().singleQuoted().line("sky"), new Scalar().singleQuoted().line("blue"))
                .entry(new Scalar().singleQuoted().line("sea"), new Scalar().singleQuoted().line("green"))
            );
        }
    }

    @Nested class givenBlockMappingWithSpacesInSingleQuotedKeyAndValue extends SingleDocument {
        @BeforeEach void setup() {
            input = "' sky high ': ' light blue '\n' sea deep ': ' dark green '";
            expected = new Document().node(new Mapping()
                .entry(new Scalar().singleQuoted().line(" sky high "), new Scalar().singleQuoted().line(" light blue "))
                .entry(new Scalar().singleQuoted().line(" sea deep "), new Scalar().singleQuoted().line(" dark green "))
            );
        }
    }

    @Nested class givenBlockMappingWithMarkedKeys extends SingleDocument {
        @BeforeEach void setup() {
            input = "? sky: blue\n? sea: green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasMarkedKey(true).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasMarkedKey(true).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithFirstKeyMarked extends SingleDocument {
        @BeforeEach void setup() {
            input = "? sky: blue\nsea: green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasMarkedKey(true).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasMarkedKey(false).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithLastKeyMarked extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: blue\n? sea: green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasMarkedKey(false).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasMarkedKey(true).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithNewLineAfterKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:\n" +
                "  blue\n" +
                "sea:\n" +
                "  green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasNlAfterKey(true).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasNlAfterKey(true).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithNewLineAfterFirstKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:\n" +
                "  blue\n" +
                "sea: green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasNlAfterKey(true).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasNlAfterKey(false).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithNewLineAfterLastKey extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky: blue\n" +
                "sea:\n" +
                "  green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry().hasNlAfterKey(false).key(new Scalar().line("sky")).value(new Scalar().line("blue")))
                .entry(new Mapping.Entry().hasNlAfterKey(true).key(new Scalar().line("sea")).value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithIndentedKeys extends SingleDocument {
        @BeforeEach void setup() {
            input = " sky: blue\n" +
                " sea: green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry()
                    .key(new Scalar().line(new Line().indent(1).text("sky")))
                    .value(new Scalar().line("blue")))
                .entry(new Mapping.Entry()
                    .key(new Scalar().line(new Line().indent(1).text("sea")))
                    .value(new Scalar().line("green")))
            );
        }
    }

    @Nested class givenBlockMappingWithIndentedValues extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:   blue\nsea:     green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry()
                    .key(new Scalar().line("sky"))
                    .value(new Scalar().line(new Line().indent(2).text("blue"))))
                .entry(new Mapping.Entry()
                    .key(new Scalar().line("sea"))
                    .value(new Scalar().line(new Line().indent(4).text("green"))))
            );
        }
    }

    @Nested class givenBlockMappingWithIndentedKeysAndValues extends SingleDocument {
        @BeforeEach void setup() {
            input = " sky:   blue\n" +
                " sea:     green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry()
                    .key(new Scalar().line(new Line().indent(1).text("sky")))
                    .value(new Scalar().line(new Line().indent(2).text("blue"))))
                .entry(new Mapping.Entry()
                    .key(new Scalar().line(new Line().indent(1).text("sea")))
                    .value(new Scalar().line(new Line().indent(4).text("green"))))
            );
        }
    }

    @Nested class givenBlockMappingWithIndentedMarkedKeysAndValues extends SingleDocument {
        @BeforeEach void setup() {
            input = "?  sky:   blue\n?    sea:     green";
            expected = new Document().node(new Mapping()
                .entry(new Mapping.Entry()
                    .hasMarkedKey(true)
                    .key(new Scalar().line(new Line().indent(1).text("sky")))
                    .value(new Scalar().line(new Line().indent(2).text("blue"))))
                .entry(new Mapping.Entry()
                    .hasMarkedKey(true)
                    .key(new Scalar().line(new Line().indent(3).text("sea")))
                    .value(new Scalar().line(new Line().indent(4).text("green"))))
            );
        }
    }

    @Nested class givenBlockMappingToBlockMapping extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:\n" +
                "  color: blue\n" +
                "  depth: high\n" +
                "sea:\n" +
                "  color: green\n" +
                "  depth: deep";
            expected = new Document().node(new Mapping()
                .entry(new Entry()
                    .key(new Scalar().line("sky"))
                    .hasNlAfterKey(true)
                    .value(new Mapping()
                        .entry("color", "blue")
                        .entry("depth", "high")
                    ))
                .entry(new Entry()
                    .key(new Scalar().line("sea"))
                    .hasNlAfterKey(true)
                    .value(new Mapping()
                        .entry("color", "green")
                        .entry("depth", "deep")
                    ))
            );
        }
    }

    @Nested class givenBlockMappingToBlockMappingToBlockMapping extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:\n" +
                "  properties:\n" +
                "    color: blue\n" +
                "    depth: high\n" +
                "  size:\n" +
                "    width: 100%\n" +
                "sea:\n" +
                "  properties:\n" +
                "    color: green\n" +
                "    depth: deep";
            expected = new Document().node(new Mapping()
                .entry(new Entry()
                    .key(new Scalar().line("sky"))
                    .hasNlAfterKey(true)
                    .value(new Mapping()
                        .entry(new Entry()
                            .key(new Scalar().line("properties"))
                            .hasNlAfterKey(true)
                            .value(new Mapping()
                                .entry("color", "blue")
                                .entry("depth", "high")))
                        .entry(new Entry()
                            .key(new Scalar().line("size"))
                            .hasNlAfterKey(true)
                            .value(new Mapping()
                                .entry("width", "100%")
                            )
                        )
                    ))
                .entry(new Entry()
                    .key(new Scalar().line("sea"))
                    .hasNlAfterKey(true)
                    .value(new Mapping()
                        .entry(new Entry()
                            .key(new Scalar().line("properties"))
                            .hasNlAfterKey(true)
                            .value(new Mapping()
                                .entry("color", "green")
                                .entry("depth", "deep")))
                    ))
            );
        }
    }

    @Disabled
    @Nested class givenBlockMappingToSequenceOfBlockMapping extends SingleDocument {
        @BeforeEach void setup() {
            input = "sky:\n" +
                "  -\n" +
                "    color: blue\n" +
                "    depth: high\n" +
                "  -\n" +
                "    foo: bar\n" +
                "sea:\n" +
                "  color: green\n" +
                "  depth: deep";
            expected = new Document().node(new Mapping()
                .entry(new Entry()
                    .key(new Scalar().line("sky"))
                    .hasNlAfterKey(true)
                    .value(new Sequence()
                        .item(new Item()
                            .nl(true)
                            .node(new Mapping()
                                .entry("color", "blue")
                                .entry("depth", "high")))
                        .item(new Item()
                            .nl(true)
                            .node(new Mapping()
                                .entry("foo", "bar")))
                    ))
                .entry(new Entry()
                    .key(new Scalar().line("sea"))
                    .hasNlAfterKey(true)
                    .value(new Mapping()
                        .entry("color", "green")
                        .entry("depth", "deep")
                    ))
            );
        }
    }
}

File: src/test/java/test/MappingTest.kt
@file:Suppress("ClassName")

package test

import com.github.t1.yaml.model.Document
import com.github.t1.yaml.model.Mapping
import com.github.t1.yaml.model.Scalar
import com.github.t1.yaml.model.Scalar.Line
import com.github.t1.yaml.model.Sequence
import com.github.t1.yaml.model.Sequence.Item
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested

class MappingTest : AbstractYamlTest() {
    @Nested inner class givenBlockMapping : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: blue\nsea: green"
            expected = Document(node = Mapping()
                .entry("sky", "blue")
                .entry("sea", "green")
            )
        }
    }

    @Nested inner class givenBlockMappingWithSpaceInKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky high: blue\nsea deep: green"
            expected = Document(node = Mapping()
                .entry("sky high", "blue")
                .entry("sea deep", "green")
            )
        }
    }

    @Nested inner class givenBlockMappingWithColonInKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:high: blue\nsea:deep: green"
            expected = Document(node = Mapping()
                .entry("sky:high", "blue")
                .entry("sea:deep", "green")
            )
        }
    }

    @Nested inner class givenBlockMappingWithTwoLineValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: high\n" +
                "  blue\n" +
                "sea: low\n" +
                "  green"
            expected = Document(node = Mapping()
                .entry("sky", Scalar().line("high").line("blue"))
                .entry("sea", Scalar().line("low").line("green"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithDoubleQuotedKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "\"sky\": blue\n\"sea\": green"
            expected = Document(node = Mapping()
                .entry(Scalar().doubleQuoted().line("sky"), "blue")
                .entry(Scalar().doubleQuoted().line("sea"), "green")
            )
        }
    }

    @Nested inner class givenBlockMappingWithDoubleQuotedValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: \"blue\"\nsea: \"green\""
            expected = Document(node = Mapping()
                .entry("sky", Scalar().doubleQuoted().line("blue"))
                .entry("sea", Scalar().doubleQuoted().line("green"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithDoubleQuotedKeyAndValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "\"sky\": \"blue\"\n\"sea\": \"green\""
            expected = Document(node = Mapping()
                .entry(Scalar().doubleQuoted().line("sky"), Scalar().doubleQuoted().line("blue"))
                .entry(Scalar().doubleQuoted().line("sea"), Scalar().doubleQuoted().line("green"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithTwoLineSingleQuotedValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: 'high\n" +
                "blue'\n" +
                "sea: 'low\n" +
                "green'"
            expected = Document(node = Mapping()
                .entry("sky", Scalar().singleQuoted().line("high\nblue"))
                .entry("sea", Scalar().singleQuoted().line("low\ngreen"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithSingleQuotedKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "'sky': blue\n'sea': green"
            expected = Document(node = Mapping()
                .entry(Scalar().singleQuoted().line("sky"), "blue")
                .entry(Scalar().singleQuoted().line("sea"), "green")
            )
        }
    }

    @Nested inner class givenBlockMappingWithSingleQuotedValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: 'blue'\nsea: 'green'"
            expected = Document(node = Mapping()
                .entry("sky", Scalar().singleQuoted().line("blue"))
                .entry("sea", Scalar().singleQuoted().line("green"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithTwoLineDoubleQuotedValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: \"high\n" +
                "blue\"\n" +
                "sea: \"low\n" +
                "green\""
            expected = Document(node = Mapping()
                .entry("sky", Scalar().doubleQuoted().line("high\nblue"))
                .entry("sea", Scalar().doubleQuoted().line("low\ngreen"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithSingleQuotedKeyAndValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "'sky': 'blue'\n'sea': 'green'"
            expected = Document(node = Mapping()
                .entry(Scalar().singleQuoted().line("sky"), Scalar().singleQuoted().line("blue"))
                .entry(Scalar().singleQuoted().line("sea"), Scalar().singleQuoted().line("green"))
            )
        }
    }

    @Nested inner class givenBlockMappingWithSpacesInSingleQuotedKeyAndValue : SingleDocument() {
        @BeforeEach fun setup() {
            input = "' sky high ': ' light blue '\n' sea deep ': ' dark green '"
            expected = Document(node = Mapping()
                .entry(Scalar().singleQuoted().line(" sky high "), Scalar().singleQuoted().line(" light blue "))
                .entry(Scalar().singleQuoted().line(" sea deep "), Scalar().singleQuoted().line(" dark green "))
            )
        }
    }

    @Nested inner class givenBlockMappingWithMarkedKeys : SingleDocument() {
        @BeforeEach fun setup() {
            input = "? sky: blue\n? sea: green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithFirstKeyMarked : SingleDocument() {
        @BeforeEach fun setup() {
            input = "? sky: blue\nsea: green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasMarkedKey = false, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithLastKeyMarked : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: blue\n? sea: green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasMarkedKey = false, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithNewLineAfterKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:\n" +
                "  blue\n" +
                "sea:\n" +
                "  green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasNlAfterKey = true, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasNlAfterKey = true, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithNewLineAfterFirstKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:\n" +
                "  blue\n" +
                "sea: green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasNlAfterKey = true, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasNlAfterKey = false, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithNewLineAfterLastKey : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky: blue\n" +
                "sea:\n" +
                "  green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasNlAfterKey = false, key = Scalar().line("sky"), value = Scalar().line("blue")))
                .entry(Mapping.Entry(hasNlAfterKey = true, key = Scalar().line("sea"), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithIndentedKeys : SingleDocument() {
        @BeforeEach fun setup() {
            input = " sky: blue\n" + " sea: green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line(Line().indent(1).text("sky")), value = Scalar().line("blue")))
                .entry(Mapping.Entry(key = Scalar().line(Line().indent(1).text("sea")), value = Scalar().line("green")))
            )
        }
    }

    @Nested inner class givenBlockMappingWithIndentedValues : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:   blue\nsea:     green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line("sky"), value = Scalar().line(Line().indent(2).text("blue"))))
                .entry(Mapping.Entry(key = Scalar().line("sea"), value = Scalar().line(Line().indent(4).text("green"))))
            )
        }
    }

    @Nested inner class givenBlockMappingWithIndentedKeysAndValues : SingleDocument() {
        @BeforeEach fun setup() {
            input = " sky:   blue\n" + " sea:     green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line(Line().indent(1).text("sky")), value = Scalar().line(Line().indent(2).text("blue"))))
                .entry(Mapping.Entry(key = Scalar().line(Line().indent(1).text("sea")), value = Scalar().line(Line().indent(4).text("green"))))
            )
        }
    }

    @Nested inner class givenBlockMappingWithIndentedMarkedKeysAndValues : SingleDocument() {
        @BeforeEach fun setup() {
            input = "?  sky:   blue\n?    sea:     green"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line(Line().indent(1).text("sky")), value = Scalar().line(Line().indent(2).text("blue"))))
                .entry(Mapping.Entry(hasMarkedKey = true, key = Scalar().line(Line().indent(3).text("sea")), value = Scalar().line(Line().indent(4).text("green"))))
            )
        }
    }

    @Nested inner class givenBlockMappingToBlockMapping : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:\n" +
                "  color: blue\n" +
                "  depth: high\n" +
                "sea:\n" +
                "  color: green\n" +
                "  depth: deep"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line("sky"), hasNlAfterKey = true, value = Mapping()
                    .entry("color", "blue")
                    .entry("depth", "high")
                ))
                .entry(Mapping.Entry(key = Scalar().line("sea"), hasNlAfterKey = true, value = Mapping()
                    .entry("color", "green")
                    .entry("depth", "deep")
                ))
            )
        }
    }

    @Nested inner class givenBlockMappingToBlockMappingToBlockMapping : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:\n" +
                "  properties:\n" +
                "    color: blue\n" +
                "    depth: high\n" +
                "  size:\n" +
                "    width: 100%\n" +
                "sea:\n" +
                "  properties:\n" +
                "    color: green\n" +
                "    depth: deep"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line("sky"), hasNlAfterKey = true, value = Mapping()
                    .entry(Mapping.Entry(key = Scalar().line("properties"), hasNlAfterKey = true, value = Mapping()
                        .entry("color", "blue")
                        .entry("depth", "high")))
                    .entry(Mapping.Entry(key = Scalar().line("size"), hasNlAfterKey = true, value = Mapping()
                        .entry("width", "100%")
                    )
                    )
                ))
                .entry(Mapping.Entry(key = Scalar().line("sea"), hasNlAfterKey = true, value = Mapping()
                    .entry(Mapping.Entry(key = Scalar().line("properties"), hasNlAfterKey = true, value = Mapping()
                        .entry("color", "green")
                        .entry("depth", "deep")))
                ))
            )
        }
    }

    @Nested inner class givenBlockMappingToSequenceOfBlockMapping : SingleDocument() {
        @BeforeEach fun setup() {
            input = "sky:\n" +
                "  -\n" +
                "    color: blue\n" +
                "    depth: high\n" +
                "  -\n" +
                "    foo: bar\n" +
                "sea:\n" +
                "  color: green\n" +
                "  depth: deep"
            expected = Document(node = Mapping()
                .entry(Mapping.Entry(key = Scalar().line("sky"), hasNlAfterKey = true, value = Sequence()
                    .item(Item(nl = true, node = Mapping()
                        .entry("color", "blue")
                        .entry("depth", "high")))
                    .item(Item(nl = true, node = Mapping()
                        .entry("foo", "bar")))
                ))
                .entry(Mapping.Entry(key = Scalar().line("sea"), hasNlAfterKey = true, value = Mapping()
                    .entry("color", "green")
                    .entry("depth", "deep")
                ))
            )
        }
    }
}
